File: agent_planning/libraries/zerotrust-auth-core/packages/go/pkg/jwt/manager.go
// Package jwt provides JWT authentication services with Zero Trust principles.
package jwt

import (
	"context"
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"
)

// Manager handles JWT operations with Zero Trust capabilities
type Manager struct {
	keyManager *KeyManager
	blacklist  Blacklist
	config     *Config
}

// Config represents JWT configuration
type Config struct {
	Secret           string        `json:"-"` // Never serialize secrets
	ExpiryDuration   time.Duration `json:"expiry_duration"`
	RefreshDuration  time.Duration `json:"refresh_duration"`
	Issuer           string        `json:"issuer"`
	RotationDuration time.Duration `json:"rotation_duration"`
}

// Claims represents JWT claims with Zero Trust attributes
type Claims struct {
	UserID      string   `json:"user_id"`
	Email       string   `json:"email"`
	Roles       []string `json:"roles"`
	Permissions []string `json:"permissions"`
	DeviceID    string   `json:"device_id,omitempty"`
	TrustLevel  int      `json:"trust_level"`
	jwt.RegisteredClaims
}

// Token represents a complete token response
type Token struct {
	AccessToken  string    `json:"access_token"`
	RefreshToken string    `json:"refresh_token"`
	TokenType    string    `json:"token_type"`
	ExpiresAt    time.Time `json:"expires_at"`
	TrustLevel   int       `json:"trust_level"`
}

// TokenRequest represents a token generation request
type TokenRequest struct {
	UserID      string   `json:"user_id" validate:"required"`
	Email       string   `json:"email" validate:"required,email"`
	Roles       []string `json:"roles"`
	Permissions []string `json:"permissions"`
	DeviceID    string   `json:"device_id,omitempty"`
	TrustLevel  int      `json:"trust_level" validate:"min=0,max=100"`
}

// Key represents a JWT signing key with metadata
type Key struct {
	ID        string    `json:"id"`
	Key       []byte    `json:"-"` // Never serialize the actual key
	CreatedAt time.Time `json:"created_at"`
	ExpiresAt time.Time `json:"expires_at"`
	IsActive  bool      `json:"is_active"`
}

// KeyManager manages JWT signing keys with rotation support
type KeyManager struct {
	mu           sync.RWMutex
	keys         map[string]*Key
	currentKeyID string
	rotationDur  time.Duration
}

// Blacklist interface for token blacklisting
type Blacklist interface {
	Add(ctx context.Context, jti, reason string, expiresAt time.Time) error
	IsBlacklisted(ctx context.Context, tokenString string) (bool, error)
	Remove(ctx context.Context, jti string) error
	Cleanup(ctx context.Context) error
}

// Custom errors
var (
	ErrTokenBlacklisted = fmt.Errorf("token has been blacklisted")
	ErrInvalidToken     = fmt.Errorf("invalid token")
	ErrExpiredToken     = fmt.Errorf("token has expired")
	ErrTokenNotActive   = fmt.Errorf("token not yet active")
)

// NewManager creates a new JWT manager with Zero Trust capabilities
func NewManager(config *Config) (*Manager, error) {
	if err := validateConfig(config); err != nil {
		return nil, fmt.Errorf("invalid config: %w", err)
	}

	keyManager := NewKeyManager([]byte(config.Secret), config.RotationDuration)
	blacklist := NewMemoryBlacklist() // Default implementation

	return &Manager{
		keyManager: keyManager,
		blacklist:  blacklist,
		config:     config,
	}, nil
}

// validateConfig validates the JWT configuration
func validateConfig(config *Config) error {
	if config == nil {
		return fmt.Errorf("config cannot be nil")
	}
	if len(config.Secret) < 32 {
		return fmt.Errorf("JWT secret must be at least 32 characters")
	}
	if config.ExpiryDuration <= 0 {
		return fmt.Errorf("expiry duration must be positive")
	}
	if config.RefreshDuration <= 0 {
		return fmt.Errorf("refresh duration must be positive")
	}
	if config.Issuer == "" {
		return fmt.Errorf("issuer cannot be empty")
	}
	if config.RotationDuration <= 0 {
		config.RotationDuration = 24 * time.Hour // Default to 24 hours
	}
	return nil
}

// GenerateToken creates a new JWT token with trust level
func (m *Manager) GenerateToken(ctx context.Context, req *TokenRequest) (*Token, error) {
	if err := m.validateTokenRequest(req); err != nil {
		return nil, fmt.Errorf("invalid token request: %w", err)
	}

	now := time.Now()
	expiresAt := now.Add(m.config.ExpiryDuration)
	jti := uuid.New().String()

	claims := &Claims{
		UserID:      req.UserID,
		Email:       req.Email,
		Roles:       req.Roles,
		Permissions: req.Permissions,
		DeviceID:    req.DeviceID,
		TrustLevel:  req.TrustLevel,
		RegisteredClaims: jwt.RegisteredClaims{
			Issuer:    m.config.Issuer,
			Subject:   req.UserID,
			ExpiresAt: jwt.NewNumericDate(expiresAt),
			IssuedAt:  jwt.NewNumericDate(now),
			NotBefore: jwt.NewNumericDate(now),
			ID:        jti,
		},
	}

	// Get current signing key
	currentKey := m.keyManager.GetCurrentKey()
	if currentKey == nil {
		return nil, fmt.Errorf("no active signing key available")
	}

	// Create token with key ID in header
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	token.Header["kid"] = currentKey.ID

	tokenString, err := token.SignedString(currentKey.Key)
	if err != nil {
		return nil, fmt.Errorf("failed to sign token: %w", err)
	}

	// Generate refresh token
	refreshToken, err := m.generateRefreshToken(req.UserID)
	if err != nil {
		return nil, fmt.Errorf("failed to generate refresh token: %w", err)
	}

	return &Token{
		AccessToken:  tokenString,
		RefreshToken: refreshToken,
		TokenType:    "Bearer",
		ExpiresAt:    expiresAt,
		TrustLevel:   req.TrustLevel,
	}, nil
}

// validateTokenRequest validates the token generation request
func (m *Manager) validateTokenRequest(req *TokenRequest) error {
	if req == nil {
		return fmt.Errorf("token request cannot be nil")
	}
	if req.UserID == "" {
		return fmt.Errorf("user ID cannot be empty")
	}
	if req.Email == "" {
		return fmt.Errorf("email cannot be empty")
	}
	if req.TrustLevel < 0 || req.TrustLevel > 100 {
		return fmt.Errorf("trust level must be between 0 and 100")
	}
	return nil
}

// ValidateToken validates a JWT token and returns claims
func (m *Manager) ValidateToken(ctx context.Context, tokenString string) (*Claims, error) {
	if tokenString == "" {
		return nil, ErrInvalidToken
	}

	// Check blacklist first
	if blacklisted, err := m.blacklist.IsBlacklisted(ctx, tokenString); err != nil {
		return nil, fmt.Errorf("blacklist check failed: %w", err)
	} else if blacklisted {
		return nil, ErrTokenBlacklisted
	}

	token, err := jwt.ParseWithClaims(tokenString, &Claims{}, func(token *jwt.Token) (interface{}, error) {
		// Verify signing method
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}

		// Check for key ID in token header
		if kidInterface, ok := token.Header["kid"]; ok {
			if kid, ok := kidInterface.(string); ok {
				if key := m.keyManager.GetKey(kid); key != nil {
					return key.Key, nil
				}
			}
		}

		// Fallback to current key for tokens without key ID (backward compatibility)
		if currentKey := m.keyManager.GetCurrentKey(); currentKey != nil {
			return currentKey.Key, nil
		}

		return nil, fmt.Errorf("no valid signing key found")
	})

	if err != nil {
		return nil, m.parseJWTError(err)
	}

	if claims, ok := token.Claims.(*Claims); ok && token.Valid {
		return claims, nil
	}

	return nil, ErrInvalidToken
}

// parseJWTError converts JWT errors to custom errors
func (m *Manager) parseJWTError(err error) error {
	errMsg := err.Error()
	if strings.Contains(errMsg, "expired") {
		return ErrExpiredToken
	} else if strings.Contains(errMsg, "not valid yet") {
		return ErrTokenNotActive
	} else if strings.Contains(errMsg, "malformed") {
		return ErrInvalidToken
	}
	return fmt.Errorf("token validation failed: %w", err)
}

// BlacklistToken adds a token to the blacklist
func (m *Manager) BlacklistToken(ctx context.Context, tokenString, reason string) error {
	if tokenString == "" {
		return fmt.Errorf("token string cannot be empty")
	}
	if reason == "" {
		return fmt.Errorf("reason cannot be empty")
	}

	// Extract JTI and expiration from token for efficient blacklisting
	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
		// We don't validate signature here, just extract claims
		return []byte("dummy"), nil
	})

	if err != nil && !strings.Contains(err.Error(), "signature is invalid") {
		return fmt.Errorf("failed to parse token for blacklisting: %w", err)
	}

	if claims, ok := token.Claims.(jwt.MapClaims); ok {
		jti, _ := claims["jti"].(string)
		expUnix, _ := claims["exp"].(float64)
		
		if jti == "" {
			return fmt.Errorf("token missing JTI claim")
		}
		
		exp := time.Unix(int64(expUnix), 0)
		return m.blacklist.Add(ctx, jti, reason, exp)
	}

	return fmt.Errorf("failed to extract claims for blacklisting")
}

// generateRefreshToken creates a refresh token
func (m *Manager) generateRefreshToken(userID string) (string, error) {
	now := time.Now()
	expiresAt := now.Add(m.config.RefreshDuration)

	claims := jwt.MapClaims{
		"user_id": userID,
		"type":    "refresh",
		"exp":     expiresAt.Unix(),
		"iat":     now.Unix(),
		"jti":     uuid.New().String(),
	}

	currentKey := m.keyManager.GetCurrentKey()
	if currentKey == nil {
		return "", fmt.Errorf("no active signing key available")
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString(currentKey.Key)
}

// RefreshToken validates a refresh token and generates new access token
func (m *Manager) RefreshToken(ctx context.Context, refreshToken string, req *TokenRequest) (*Token, error) {
	// Validate refresh token
	token, err := jwt.Parse(refreshToken, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		
		currentKey := m.keyManager.GetCurrentKey()
		if currentKey == nil {
			return nil, fmt.Errorf("no active signing key available")
		}
		return currentKey.Key, nil
	})

	if err != nil {
		return nil, fmt.Errorf("invalid refresh token: %w", err)
	}

	if claims, ok := token.Claims.(jwt.MapClaims); ok && token.Valid {
		tokenType, _ := claims["type"].(string)
		userID, _ := claims["user_id"].(string)
		
		if tokenType != "refresh" {
			return nil, fmt.Errorf("not a refresh token")
		}
		
		if userID != req.UserID {
			return nil, fmt.Errorf("user ID mismatch")
		}

		// Generate new access token
		return m.GenerateToken(ctx, req)
	}

	return nil, fmt.Errorf("invalid refresh token claims")
}

// SetBlacklist sets the blacklist implementation
func (m *Manager) SetBlacklist(blacklist Blacklist) {
	m.blacklist = blacklist
}

// GetConfig returns a copy of the configuration (without secrets)
func (m *Manager) GetConfig() Config {
	return Config{
		ExpiryDuration:   m.config.ExpiryDuration,
		RefreshDuration:  m.config.RefreshDuration,
		Issuer:           m.config.Issuer,
		RotationDuration: m.config.RotationDuration,
	}
}

// KeyManager Implementation

// NewKeyManager creates a new key manager with rotation support
func NewKeyManager(initialSecret []byte, rotationDur time.Duration) *KeyManager {
	keyID := generateKeyID()
	now := time.Now()

	key := &Key{
		ID:        keyID,
		Key:       initialSecret,
		CreatedAt: now,
		ExpiresAt: now.Add(rotationDur * 2), // Allow overlap for token validation
		IsActive:  true,
	}

	return &KeyManager{
		keys:         map[string]*Key{keyID: key},
		currentKeyID: keyID,
		rotationDur:  rotationDur,
	}
}

// generateKeyID creates a unique key identifier
func generateKeyID() string {
	bytes := make([]byte, 16)
	rand.Read(bytes)
	return base64.URLEncoding.EncodeToString(bytes)
}

// GetCurrentKey returns the current active signing key
func (km *KeyManager) GetCurrentKey() *Key {
	km.mu.RLock()
	defer km.mu.RUnlock()

	if key, exists := km.keys[km.currentKeyID]; exists && key.IsActive {
		return key
	}
	return nil
}

// GetKey returns a specific key by ID for token validation
func (km *KeyManager) GetKey(keyID string) *Key {
	km.mu.RLock()
	defer km.mu.RUnlock()

	return km.keys[keyID]
}

// RotateKey creates a new signing key and marks it as current
func (km *KeyManager) RotateKey() error {
	km.mu.Lock()
	defer km.mu.Unlock()

	// Generate new key
	newKeyBytes := make([]byte, 32)
	if _, err := rand.Read(newKeyBytes); err != nil {
		return fmt.Errorf("failed to generate new key: %w", err)
	}

	newKeyID := generateKeyID()
	now := time.Now()

	newKey := &Key{
		ID:        newKeyID,
		Key:       newKeyBytes,
		CreatedAt: now,
		ExpiresAt: now.Add(km.rotationDur * 2), // Allow overlap for token validation
		IsActive:  true,
	}

	// Mark current key as inactive
	if currentKey, exists := km.keys[km.currentKeyID]; exists {
		currentKey.IsActive = false
	}

	// Add new key and update current
	km.keys[newKeyID] = newKey
	km.currentKeyID = newKeyID

	// Clean up expired keys
	km.cleanupExpiredKeys()

	return nil
}

// cleanupExpiredKeys removes keys that are past their expiration time
func (km *KeyManager) cleanupExpiredKeys() {
	now := time.Now()
	for keyID, key := range km.keys {
		if now.After(key.ExpiresAt) && keyID != km.currentKeyID {
			delete(km.keys, keyID)
		}
	}
}

// GetStats returns statistics about the key manager
func (km *KeyManager) GetStats() map[string]interface{} {
	km.mu.RLock()
	defer km.mu.RUnlock()

	activeKeys := 0
	expiredKeys := 0
	now := time.Now()

	for _, key := range km.keys {
		if key.IsActive && now.Before(key.ExpiresAt) {
			activeKeys++
		} else if now.After(key.ExpiresAt) {
			expiredKeys++
		}
	}

	return map[string]interface{}{
		"total_keys":      len(km.keys),
		"active_keys":     activeKeys,
		"expired_keys":    expiredKeys,
		"current_key_id":  km.currentKeyID,
		"rotation_period": km.rotationDur.String(),
	}
}

File: agent_planning/libraries/zerotrust-auth-core/packages/typescript/src/jwt/manager.ts
/**
 * JWT authentication services with Zero Trust principles
 */

import jwt from 'jsonwebtoken';
import { v4 as uuidv4 } from 'uuid';
import { Blacklist, MemoryBlacklist } from '../blacklist/blacklist.js';
import { TrustLevel } from '../trust/calculator.js';

/**
 * JWT Manager configuration
 */
export interface JWTConfig {
  secret: string;
  expiryDuration: number; // milliseconds
  refreshDuration: number; // milliseconds
  issuer: string;
  rotationDuration?: number; // milliseconds
}

/**
 * JWT claims with Zero Trust attributes
 */
export interface JWTClaims {
  userId: string;
  email: string;
  roles: string[];
  permissions: string[];
  deviceId?: string;
  trustLevel: number;
  iat: number;
  exp: number;
  iss: string;
  sub: string;
  jti: string;
  nbf?: number;
}

/**
 * Token response
 */
export interface Token {
  accessToken: string;
  refreshToken: string;
  tokenType: string;
  expiresAt: Date;
  trustLevel: number;
}

/**
 * Token generation request
 */
export interface TokenRequest {
  userId: string;
  email: string;
  roles: string[];
  permissions: string[];
  deviceId?: string;
  trustLevel: number;
}

/**
 * JWT signing key with metadata
 */
export interface JWTKey {
  id: string;
  key: string;
  createdAt: Date;
  expiresAt: Date;
  isActive: boolean;
}

/**
 * Custom JWT errors
 */
export class JWTError extends Error {
  constructor(message: string, public code: string) {
    super(message);
    this.name = 'JWTError';
  }
}

export class TokenBlacklistedError extends JWTError {
  constructor() {
    super('Token has been blacklisted', 'TOKEN_BLACKLISTED');
  }
}

export class InvalidTokenError extends JWTError {
  constructor(message = 'Invalid token') {
    super(message, 'INVALID_TOKEN');
  }
}

export class ExpiredTokenError extends JWTError {
  constructor() {
    super('Token has expired', 'EXPIRED_TOKEN');
  }
}

export class TokenNotActiveError extends JWTError {
  constructor() {
    super('Token not yet active', 'TOKEN_NOT_ACTIVE');
  }
}

/**
 * Key manager for JWT signing keys with rotation support
 */
export class KeyManager {
  private keys = new Map<string, JWTKey>();
  private currentKeyId: string;
  private rotationDuration: number;

  constructor(initialSecret: string, rotationDuration: number = 24 * 60 * 60 * 1000) {
    this.rotationDuration = rotationDuration;
    
    const keyId = this.generateKeyId();
    const now = new Date();
    
    const key: JWTKey = {
      id: keyId,
      key: initialSecret,
      createdAt: now,
      expiresAt: new Date(now.getTime() + rotationDuration * 2), // Allow overlap
      isActive: true
    };

    this.keys.set(keyId, key);
    this.currentKeyId = keyId;
  }

  /**
   * Get the current active signing key
   */
  getCurrentKey(): JWTKey | null {
    const key = this.keys.get(this.currentKeyId);
    return key && key.isActive ? key : null;
  }

  /**
   * Get a specific key by ID for token validation
   */
  getKey(keyId: string): JWTKey | null {
    return this.keys.get(keyId) || null;
  }

  /**
   * Rotate the signing key
   */
  rotateKey(): void {
    // Generate new key
    const newKeyBytes = this.generateSecureKey();
    const newKeyId = this.generateKeyId();
    const now = new Date();

    const newKey: JWTKey = {
      id: newKeyId,
      key: newKeyBytes,
      createdAt: now,
      expiresAt: new Date(now.getTime() + this.rotationDuration * 2),
      isActive: true
    };

    // Mark current key as inactive
    const currentKey = this.keys.get(this.currentKeyId);
    if (currentKey) {
      currentKey.isActive = false;
    }

    // Add new key and update current
    this.keys.set(newKeyId, newKey);
    this.currentKeyId = newKeyId;

    // Clean up expired keys
    this.cleanupExpiredKeys();
  }

  /**
   * Get key manager statistics
   */
  getStats(): Record<string, any> {
    const now = new Date();
    let activeKeys = 0;
    let expiredKeys = 0;

    for (const key of this.keys.values()) {
      if (key.isActive && now < key.expiresAt) {
        activeKeys++;
      } else if (now > key.expiresAt) {
        expiredKeys++;
      }
    }

    return {
      totalKeys: this.keys.size,
      activeKeys,
      expiredKeys,
      currentKeyId: this.currentKeyId,
      rotationPeriod: `${this.rotationDuration / 1000}s`
    };
  }

  /**
   * Generate a unique key identifier
   */
  private generateKeyId(): string {
    return Buffer.from(uuidv4().replace(/-/g, '')).toString('base64').substring(0, 16);
  }

  /**
   * Generate a secure key
   */
  private generateSecureKey(): string {
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*';
    let result = '';
    for (let i = 0; i < 64; i++) {
      result += chars.charAt(Math.floor(Math.random() * chars.length));
    }
    return result;
  }

  /**
   * Clean up expired keys
   */
  private cleanupExpiredKeys(): void {
    const now = new Date();
    for (const [keyId, key] of this.keys) {
      if (now > key.expiresAt && keyId !== this.currentKeyId) {
        this.keys.delete(keyId);
      }
    }
  }
}

/**
 * JWT Manager with Zero Trust capabilities
 */
export class JWTManager {
  private keyManager: KeyManager;
  private blacklist: Blacklist;
  private config: JWTConfig;

  constructor(config: JWTConfig) {
    this.validateConfig(config);
    this.config = config;
    this.keyManager = new KeyManager(config.secret, config.rotationDuration);
    this.blacklist = new MemoryBlacklist(); // Default implementation
  }

  /**
   * Generate a new JWT token with trust level
   */
  async generateToken(request: TokenRequest): Promise<Token> {
    this.validateTokenRequest(request);

    const now = Math.floor(Date.now() / 1000);
    const expiresAt = new Date(Date.now() + this.config.expiryDuration);
    const jti = uuidv4();

    const claims: JWTClaims = {
      userId: request.userId,
      email: request.email,
      roles: request.roles,
      permissions: request.permissions,
      deviceId: request.deviceId,
      trustLevel: request.trustLevel,
      iat: now,
      exp: Math.floor(expiresAt.getTime() / 1000),
      iss: this.config.issuer,
      sub: request.userId,
      jti,
      nbf: now
    };

    // Get current signing key
    const currentKey = this.keyManager.getCurrentKey();
    if (!currentKey) {
      throw new Error('No active signing key available');
    }

    // Create token with key ID in header
    const token = jwt.sign(claims, currentKey.key, {
      algorithm: 'HS256',
      header: { kid: currentKey.id }
    });

    // Generate refresh token
    const refreshToken = await this.generateRefreshToken(request.userId);

    return {
      accessToken: token,
      refreshToken,
      tokenType: 'Bearer',
      expiresAt,
      trustLevel: request.trustLevel
    };
  }

  /**
   * Validate a JWT token and return claims
   */
  async validateToken(tokenString: string): Promise<JWTClaims> {
    if (!tokenString) {
      throw new InvalidTokenError();
    }

    // Check blacklist first
    const blacklisted = await this.blacklist.isBlacklisted(tokenString);
    if (blacklisted) {
      throw new TokenBlacklistedError();
    }

    try {
      const decoded = jwt.verify(tokenString, (header, callback) => {
        // Check for key ID in token header
        if (header.kid) {
          const key = this.keyManager.getKey(header.kid);
          if (key) {
            callback(null, key.key);
            return;
          }
        }

        // Fallback to current key for tokens without key ID
        const currentKey = this.keyManager.getCurrentKey();
        if (currentKey) {
          callback(null, currentKey.key);
          return;
        }

        callback(new Error('No valid signing key found'));
      }, {
        algorithms: ['HS256'],
        issuer: this.config.issuer
      }) as JWTClaims;

      return decoded;
    } catch (error: any) {
      throw this.parseJWTError(error);
    }
  }

  /**
   * Blacklist a token
   */
  async blacklistToken(tokenString: string, reason: string): Promise<void> {
    if (!tokenString) {
      throw new Error('Token string cannot be empty');
    }
    if (!reason) {
      throw new Error('Reason cannot be empty');
    }

    try {
      // Extract JTI and expiration from token for efficient blacklisting
      const decoded = jwt.decode(tokenString, { complete: true });
      
      if (!decoded || typeof decoded === 'string') {
        throw new Error('Invalid token format');
      }

      const payload = decoded.payload as any;
      const jti = payload.jti;
      const exp = payload.exp;

      if (!jti) {
        throw new Error('Token missing JTI claim');
      }

      const expiresAt = new Date(exp * 1000);
      await this.blacklist.add(jti, reason, expiresAt);
    } catch (error) {
      throw new Error(`Failed to blacklist token: ${error}`);
    }
  }

  /**
   * Refresh a token using a valid refresh token
   */
  async refreshToken(refreshToken: string, request: TokenRequest): Promise<Token> {
    try {
      const currentKey = this.keyManager.getCurrentKey();
      if (!currentKey) {
        throw new Error('No active signing key available');
      }

      const decoded = jwt.verify(refreshToken, currentKey.key, {
        algorithms: ['HS256']
      }) as any;

      if (decoded.type !== 'refresh') {
        throw new Error('Not a refresh token');
      }

      if (decoded.user_id !== request.userId) {
        throw new Error('User ID mismatch');
      }

      // Generate new access token
      return await this.generateToken(request);
    } catch (error) {
      throw new Error(`Invalid refresh token: ${error}`);
    }
  }

  /**
   * Set the blacklist implementation
   */
  setBlacklist(blacklist: Blacklist): void {
    this.blacklist = blacklist;
  }

  /**
   * Get configuration (without secrets)
   */
  getConfig(): Omit<JWTConfig, 'secret'> {
    const { secret, ...config } = this.config;
    return config;
  }

  /**
   * Validate JWT configuration
   */
  private validateConfig(config: JWTConfig): void {
    if (!config) {
      throw new Error('Config cannot be null');
    }
    if (config.secret.length < 32) {
      throw new Error('JWT secret must be at least 32 characters');
    }
    if (config.expiryDuration <= 0) {
      throw new Error('Expiry duration must be positive');
    }
    if (config.refreshDuration <= 0) {
      throw new Error('Refresh duration must be positive');
    }
    if (!config.issuer) {
      throw new Error('Issuer cannot be empty');
    }
    if (!config.rotationDuration) {
      config.rotationDuration = 24 * 60 * 60 * 1000; // Default to 24 hours
    }
  }

  /**
   * Validate token generation request
   */
  private validateTokenRequest(request: TokenRequest): void {
    if (!request) {
      throw new Error('Token request cannot be null');
    }
    if (!request.userId) {
      throw new Error('User ID cannot be empty');
    }
    if (!request.email) {
      throw new Error('Email cannot be empty');
    }
    if (request.trustLevel < 0 || request.trustLevel > 100) {
      throw new Error('Trust level must be between 0 and 100');
    }
  }

  /**
   * Generate a refresh token
   */
  private async generateRefreshToken(userId: string): Promise<string> {
    const now = Math.floor(Date.now() / 1000);
    const expiresAt = Math.floor((Date.now() + this.config.refreshDuration) / 1000);

    const claims = {
      user_id: userId,
      type: 'refresh',
      exp: expiresAt,
      iat: now,
      jti: uuidv4()
    };

    const currentKey = this.keyManager.getCurrentKey();
    if (!currentKey) {
      throw new Error('No active signing key available');
    }

    return jwt.sign(claims, currentKey.key, { algorithm: 'HS256' });
  }

  /**
   * Parse JWT errors to custom errors
   */
  private parseJWTError(error: any): Error {
    const errMsg = error.message || '';
    
    if (errMsg.includes('expired')) {
      return new ExpiredTokenError();
    } else if (errMsg.includes('not active')) {
      return new TokenNotActiveError();
    } else if (errMsg.includes('malformed') || errMsg.includes('invalid')) {
      return new InvalidTokenError();
    }
    
    return new InvalidTokenError(`Token validation failed: ${errMsg}`);
  }
}